pyccl/nl_pt/power.py

```diff
@@ -0,0 +1,583 @@
+                redshifts.
+            c2 (array_like): 2-nd order bias for the IA
+                tracer being correlated at the same set of input
+                redshifts.
+            cd (array_like): overdensity bias for the IA
+                tracer being correlated at the same set of input
+                redshifts.
+
+        Returns:
+            array_like: 2D array of shape `(N_k, N_z)`, where `N_k` \
+                is the size of this object's `ks` attribute, and \
+                `N_z` is the size of the input redshift-dependent \
+                biases and growth factor.
+        """
+        a00e, c00e, a0e0e, a0b0b = self.ia_ta
+        a0e2, b0e2, d0ee2, d0bb2 = self.ia_mix
+
+        pim = (c1[None, :] * Pd1d1 +
+               (g4*cd)[None, :] * (a00e + c00e)[:, None] +
+               (g4*c2)[None, :] * (a0e2 + b0e2)[:, None])
+        return pim
+
+    def get_pmm(self, Pd1d1_lin, g4):
+        """ Get the one-loop matter power spectrum.
+
+        Args:
+            Pd1d1_lin (array_like): 1-loop linear matter power spectrum
+                at the wavenumber values given by this object's
+                `ks` list.
+            g4 (array_like): fourth power of the growth factor at
+                a number of redshifts.
+
+        Returns:
+            array_like: 2D array of shape `(N_k, N_z)`, where `N_k` \
+                is the size of this object's `ks` attribute, and \
+                `N_z` is the size of the input redshift-dependent \
+                biases and growth factor.
+        """
+        P1loop = g4[None, :] * self.one_loop_dd[0][:, None]
+        pmm = (Pd1d1_lin + P1loop)
+        return pmm
+
+
+def get_pt_pk2d(cosmo, tracer1, tracer2=None, ptc=None,
+                sub_lowk=False, nonlin_pk_type='halofit',
+                a_arr=None, extrap_order_lok=1, extrap_order_hik=2,
+                return_ia_bb=False, return_ia_ee_and_bb=False):
+    """Returns a :class:`~pyccl.pk2d.Pk2D` object containing
+    the PT power spectrum for two quantities defined by
+    two :class:`~pyccl.nl_pt.tracers.PTTracer` objects.
+
+    Args:
+        cosmo (:class:`~pyccl.core.Cosmology`): a Cosmology object.
+        tracer1 (:class:`~pyccl.nl_pt.tracers.PTTracer`): the first
+            tracer being correlated.
+        ptc (:class:`PTCalculator`): a perturbation theory
+            calculator.
+        tracer2 (:class:`~pyccl.nl_pt.tracers.PTTracer`): the second
+            tracer being correlated. If `None`, the auto-correlation
+            of the first tracer will be returned.
+        sub_lowk (bool): if True, the small-scale white noise
+            contribution will be subtracted for number counts
+            auto-correlations.
+        nonlin_pk_type (str): type of 1-loop matter power spectrum
+            to use. 'linear' for linear P(k), 'halofit' for halofit
+            non-linear power spectrum, 'spt' for standard perturbation
```

Comment:
halofit as the option here is not accurate. The non-linear power could be an emulator. The options should be linear, non-linear or spt

Reply:
Done